### PR TITLE
Add flake8 and pydocstyle lints

### PR DESCRIPTION
**Pull request by seocam**
Addresses #241 (but does not close)

### CODE REVIEW

File: plugins/module_utils/ansible_freeipa_module.py [hunk 1]
```diff
@@ -338,9 +324,7 @@ def is_valid_port(port):
 
 
 def is_ipv4_addr(ipaddr):
-    """
-    Test if figen IP address is a valid IPv4 address
-    """
+    """Test if figen IP address is a valid IPv4 address."""
```
Comment (t-woerner) [inline plugins/module_utils/ansible_freeipa_module.py:327]: "figen" should be "given"

File: plugins/module_utils/ansible_freeipa_module.py [hunk 2]
```diff
@@ -349,9 +333,7 @@ def is_ipv4_addr(ipaddr):
 
 
 def is_ipv6_addr(ipaddr):
-    """
-    Test if figen IP address is a valid IPv6 address
-    """
+    """Test if figen IP address is a valid IPv6 address."""
```
Comment (t-woerner) [inline plugins/module_utils/ansible_freeipa_module.py:336]: Same here.

**Comment by seocam** [comment]
> I am not convinced that the black formatting is the best here as it adds hundreds of extra lines. The code is not better readable in my opinion. Screens are a lot wider than high.

We can configure how wide we want. By default it uses 88 chars and line length I changed it to 79 to match with flake8 default. Right now anything under 160 cpl (chars per line) is valid for galaxy.

Since I code using vim with 4 terminals / screen 88 chars looks pretty good.

Leaving our 2 personal preferences aside, there are studies that points that it's easier to spot a word when reading in longer lines but shorter lines should be better for understanding.

File: plugins/modules/ipagroup.py [hunk 3]
```diff
@@ -243,28 +253,35 @@ def main():
...
         if action == "member":
-            invalid = ["description", "gid", "nonposix", "external",
-                       "nomembers"]
+            invalid = [
+                "description",
+                "gid",
+                "nonposix",
+                "external",
+                "nomembers",
+            ]
```
Comment (rjeffman) [inline plugins/modules/ipagroup.py:265]: This is one type of change that I do not like. It make elements disperse, and they don't seem to be part of the same thing.

**Comment by seocam** [comment]
While it seems that we don't actually have an agreement about black nothing was said regarding pydocstyle and flake8. I suppose that's good :)

If I redo this PR without black and open a different one for us to discuss it would that be ok for you @rjeffman and @t-woerner?

**Comment by rjeffman** [comment]
I like flake8, but we might want some exceptions (maybe a .flake8 file), since we have some rules that are not really PEP8 (like import being the first thing on files).

I never used pydocstyle, and while it didn't do anything that I dislike in your patch, I don't know how it works. I once used flake8 also on docstrings, the good thing is that it one less tool to depend on, but as I never used pydocstyle before, I don't know what we would miss by choosing flake8.

**Comment by seocam** [comment]
I've configured flake8 to match galaxy rules.

Flake8 verify a few things on docstrings but doesn't enforce the PEP257. In fact there are not many changes in docs by enforcing it. On having one for tool, it shouldn't be a big deal since this will run on CI an takes only a few seconds (not more than 10 sec).

**Comment by rjeffman** [comment]
I'm not against a second tool, if using a single one is not "good enough". If flake8 doesn't cover it all, we can use of pydocstyle.

**Comment by seocam** [comment]
Updated the PR without using black.

File: roles/ipaclient/library/ipaclient_get_otp.py [hunk 4]
```diff
@@ -189,7 +188,7 @@ def get_module_host(module):
 
 def ensure_host_present(module, api, ipahost):
     """
-    Ensures that the host exists in IPA and has the same attributes.
+    Ensure that the host exists in IPA and has the same attributes.
```
Comment (rjeffman) [inline roles/ipaclient/library/ipaclient_get_otp.py:191]: I'd remove "that the".

File: roles/ipaclient/library/ipaclient_get_otp.py [hunk 5]
```diff
@@ -246,7 +245,7 @@ def ensure_host_present(module, api, ipahost):
 
 def ensure_host_absent(module, api, host):
     """
-    Ensures that the host does not exist in IPA
+    Ensure that the host does not exist in IPA.
```
Comment (rjeffman) [inline roles/ipaclient/library/ipaclient_get_otp.py:248]: I'd remove "that the".

File: roles/ipaclient/library/ipaclient_get_otp.py [hunk 6]
```diff
@@ -135,8 +135,7 @@
 
 def get_host_diff(ipa_host, module_host):
     """
-    Compares two dictionaries containing host attributes and builds a dict
-    of differences.
+    Compare two host dicts and builds a dict of differences.
```
Comment (rjeffman) [inline roles/ipaclient/library/ipaclient_get_otp.py:138]: I think using "Build a dictionary with the differences from two host dictionaries." still shows what is being done, and doesn't look like the function is doing two different things.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 7]
```diff
@@ -157,28 +149,24 @@ def api_connect(context=None):
...
-    """
+    """Call ipa.Command without a name."""
     return api.Command[command](**args)
 
 
 def api_check_param(command, name):
-    """
-    Return if param exists in command param list
-    """
+    """Return if param exists in command param list."""
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py:162]: I'd use "Check" (or something similar) instead of "Return".

File: plugins/module_utils/ansible_freeipa_module.py [hunk 8]
```diff
@@ -123,19 +119,15 @@ def temp_kinit(principal, password):
...
         run([paths.KDESTROY, '-c', ccache_name], raiseonerr=False)
     if ccache_dir is not None:
         shutil.rmtree(ccache_dir, ignore_errors=True)
 
 
 def api_connect(context=None):
-    """
-    Create environment, initialize api and connect to ldap2
-    """
+    """Create environment, initialize api and connect to ldap2."""
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py:130]: I'd rephrase it to:
```python
"""
Initialize IPA API with the provided context.

`context` can be any of  `server`, `ansible-freeipa` and `cli_installer`, default to `server`.
"""
```

BTW, why isn't `server` the default value instead of `None`?

File: plugins/module_utils/ansible_freeipa_module.py [hunk 9]
```diff
@@ -63,9 +63,7 @@
 
 
 def valid_creds(module, principal):  # noqa
-    """
-    Get valid credintials matching the princial, try GSSAPI first
-    """
+    """Get valid credintials matching the princial, try GSSAPI first."""
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py:66]: There is a typo in `credentials`.

File: plugins/module_utils/ansible_freeipa_module.py (hunk 8 above)
Comment (seocam) [reply plugins/module_utils/ansible_freeipa_module.py:130]: I don't actually know that answer :)
Maybe we should open an issue for that?

File: roles/ipareplica/files/py3test.py [hunk 10]
```diff
@@ -2,9 +2,13 @@
 
 # Test ipaerver python3 binding
 try:
-    from ipaserver.install.server.replicainstall import install_check
+    from ipaserver.install.server.replicainstall import (  # noqa: F401
```
Comment (t-woerner) [inline roles/ipareplica/files/py3test.py:5]: Here you still have the black changes.
We can remove `py3test.py ` from all roles as the files are not used any more.

Comment (seocam) [reply roles/ipareplica/files/py3test.py:5]: Those are not black changes. Those are mine in order to add the noqa and don't break the line length.

Should I remove in this PR? I can add a commit just for that.

Comment (t-woerner) [reply roles/ipareplica/files/py3test.py:5]: a separate PR would be good - also the commented out code in the roles need to be removed.
